From 87126390d0af33cd3e3a5f671e3336ff0c90e6e2 Mon Sep 17 00:00:00 2001
From: SpiritCroc <dev@spiritcroc.de>
Date: Fri, 4 Aug 2023 10:49:56 +0200
Subject: [PATCH] Fix possible NPE in StatusBarIconList causing bootloop

08-04 10:41:28.265  4198  4198 E AndroidRuntime: FATAL EXCEPTION: main
08-04 10:41:28.265  4198  4198 E AndroidRuntime: Process: com.android.systemui, PID: 4198
08-04 10:41:28.265  4198  4198 E AndroidRuntime: java.lang.RuntimeException: Unable to create service com.android.systemui.SystemUIService: java.lang.NullPointerException: Attempt to invoke virtual method 'boolean java.lang.String.equals(java.lang.Object)' on a null object reference
08-04 10:41:28.265  4198  4198 E AndroidRuntime: 	at android.app.ActivityThread.handleCreateService(ActivityThread.java:4498)
08-04 10:41:28.265  4198  4198 E AndroidRuntime: 	at android.app.ActivityThread.-$$Nest$mhandleCreateService(Unknown Source:0)
08-04 10:41:28.265  4198  4198 E AndroidRuntime: 	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2160)
08-04 10:41:28.265  4198  4198 E AndroidRuntime: 	at android.os.Handler.dispatchMessage(Handler.java:106)
08-04 10:41:28.265  4198  4198 E AndroidRuntime: 	at android.os.Looper.loopOnce(Looper.java:201)
08-04 10:41:28.265  4198  4198 E AndroidRuntime: 	at android.os.Looper.loop(Looper.java:288)
08-04 10:41:28.265  4198  4198 E AndroidRuntime: 	at android.app.ActivityThread.main(ActivityThread.java:7918)
08-04 10:41:28.265  4198  4198 E AndroidRuntime: 	at java.lang.reflect.Method.invoke(Native Method)
08-04 10:41:28.265  4198  4198 E AndroidRuntime: 	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:548)
08-04 10:41:28.265  4198  4198 E AndroidRuntime: 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:936)
08-04 10:41:28.265  4198  4198 E AndroidRuntime: Caused by: java.lang.NullPointerException: Attempt to invoke virtual method 'boolean java.lang.String.equals(java.lang.Object)' on a null object reference
08-04 10:41:28.265  4198  4198 E AndroidRuntime: 	at com.android.systemui.statusbar.phone.StatusBarIconList.findOrInsertSlot(StatusBarIconList.java:20)
08-04 10:41:28.265  4198  4198 E AndroidRuntime: 	at com.android.systemui.statusbar.phone.StatusBarIconList.getSlot(StatusBarIconList.java:2)
08-04 10:41:28.265  4198  4198 E AndroidRuntime: 	at com.android.systemui.statusbar.phone.StatusBarIconControllerImpl.removeAllIconsForSlot(StatusBarIconControllerImpl.java:2)
08-04 10:41:28.265  4198  4198 E AndroidRuntime: 	at com.android.systemui.statusbar.phone.StatusBarIconControllerImpl.onTuningChanged(StatusBarIconControllerImpl.java:75)
08-04 10:41:28.265  4198  4198 E AndroidRuntime: 	at com.android.systemui.tuner.TunerServiceImpl.addTunable(TunerServiceImpl.java:134)
08-04 10:41:28.265  4198  4198 E AndroidRuntime: 	at com.android.systemui.statusbar.phone.StatusBarIconControllerImpl.<init>(StatusBarIconControllerImpl.java:40)
08-04 10:41:28.265  4198  4198 E AndroidRuntime: 	at com.android.systemui.statusbar.phone.StatusBarIconControllerImpl_Factory.get(StatusBarIconControllerImpl_Factory.java:93)
08-04 10:41:28.265  4198  4198 E AndroidRuntime: 	at dagger.internal.DoubleCheck.get(DoubleCheck.java:13)
08-04 10:41:28.265  4198  4198 E AndroidRuntime: 	at com.android.systemui.statusbar.phone.StatusBarSignalPolicy_Factory.get(StatusBarSignalPolicy_Factory.java:17)
08-04 10:41:28.265  4198  4198 E AndroidRuntime: 	at dagger.internal.DoubleCheck.get(DoubleCheck.java:13)
08-04 10:41:28.265  4198  4198 E AndroidRuntime: 	at com.android.systemui.statusbar.phone.CentralSurfacesImpl_Factory.get(CentralSurfacesImpl_Factory.java:76)
08-04 10:41:28.265  4198  4198 E AndroidRuntime: 	at dagger.internal.DoubleCheck.get(DoubleCheck.java:13)
08-04 10:41:28.265  4198  4198 E AndroidRuntime: 	at dagger.internal.DelegateFactory.get(DelegateFactory.java:4)
08-04 10:41:28.265  4198  4198 E AndroidRuntime: 	at com.android.systemui.dagger.DaggerReferenceGlobalRootComponent$PresentJdkOptionalInstanceProvider.get(DaggerReferenceGlobalRootComponent.java:2)
08-04 10:41:28.265  4198  4198 E AndroidRuntime: 	at com.android.systemui.dagger.DaggerReferenceGlobalRootComponent$PresentJdkOptionalInstanceProvider.get(DaggerReferenceGlobalRootComponent.java:1)
08-04 10:41:28.265  4198  4198 E AndroidRuntime: 	at dagger.internal.DelegateFactory.get(DelegateFactory.java:4)
08-04 10:41:28.265  4198  4198 E AndroidRuntime: 	at dagger.internal.DoubleCheck.get(DoubleCheck.java:13)
08-04 10:41:28.265  4198  4198 E AndroidRuntime: 	at com.android.systemui.accessibility.SystemActions.registerActions(SystemActions.java:161)
08-04 10:41:28.265  4198  4198 E AndroidRuntime: 	at com.android.systemui.accessibility.SystemActions.start(SystemActions.java:118)
08-04 10:41:28.265  4198  4198 E AndroidRuntime: 	at com.android.systemui.SystemUIApplication.startStartable(SystemUIApplication.java:36)
08-04 10:41:28.265  4198  4198 E AndroidRuntime: 	at com.android.systemui.SystemUIApplication$$ExternalSyntheticLambda3.run(R8$$SyntheticClass:58)
08-04 10:41:28.265  4198  4198 E AndroidRuntime: 	at com.android.systemui.SystemUIApplication.timeInitialization(SystemUIApplication.java:27)
08-04 10:41:28.265  4198  4198 E AndroidRuntime: 	at com.android.systemui.SystemUIApplication.startServicesIfNeeded(SystemUIApplication.java:20)
08-04 10:41:28.265  4198  4198 E AndroidRuntime: 	at com.android.systemui.SystemUIApplication.startServicesIfNeeded(SystemUIApplication.java:6)
08-04 10:41:28.265  4198  4198 E AndroidRuntime: 	at com.android.systemui.SystemUIService.onCreate(SystemUIService.java:9)
08-04 10:41:28.265  4198  4198 E AndroidRuntime: 	at android.app.ActivityThread.handleCreateService(ActivityThread.java:4485)
08-04 10:41:28.265  4198  4198 E AndroidRuntime: 	... 9 more

Change-Id: I64b1a636e79cff9f8a53d23c632fefc569ca2bdf
---
 .../android/systemui/statusbar/phone/StatusBarIconList.java    | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/packages/SystemUI/src/com/android/systemui/statusbar/phone/StatusBarIconList.java b/packages/SystemUI/src/com/android/systemui/statusbar/phone/StatusBarIconList.java
index 565481a20d95..bbe89b33b834 100644
--- a/packages/SystemUI/src/com/android/systemui/statusbar/phone/StatusBarIconList.java
+++ b/packages/SystemUI/src/com/android/systemui/statusbar/phone/StatusBarIconList.java
@@ -117,7 +117,8 @@ public class StatusBarIconList {
         final int N = mSlots.size();
         for (int i = 0; i < N; i++) {
             Slot item = mSlots.get(i);
-            if (item.getName().equals(slot)) {
+            String name = item.getName();
+            if (name != null && name.equals(slot)) {
                 return i;
             }
         }
-- 
2.43.0

